p2p/src/lib.rs

```diff
@@ -117,24 +184,90 @@ where
         event: ConnectivityEvent<NetworkingBackend>,
     ) -> error::Result<()> {
         match event {
-            ConnectivityEvent::Accept(socket) => self.create_peer(socket, PeerRole::Inbound),
-            ConnectivityEvent::Connect(address) => self
-                .network
-                .connect(address)
-                .await
-                .map(|socket| self.create_peer(socket, PeerRole::Outbound))?,
+            ConnectivityEvent::Accept(peer_id, socket) => {
+                self.create_peer(peer_id, socket, PeerRole::Inbound)
+            }
+            ConnectivityEvent::Connect(address) => {
+                self.network.connect(address).await.map(|(peer_id, socket)| {
+                    self.create_peer(peer_id, socket, PeerRole::Outbound)
+                })?
+            }
         }
 
         Ok(())
     }
 
-    fn peer_discovered(&mut self, peers: &[NetworkingBackend::Address]) -> error::Result<()> {
-        println!("peers discovered: {:#?}", peers);
+    /// Try to establish new outbound connections if the total number of
+    /// active connections the local node has is below threshold
+    ///
+    // TODO: move all peer management to separate file
+    async fn auto_connect(&mut self) -> error::Result<()> {
+        // we have enough active connections
+        if self.peers.len() >= MAX_ACTIVE_CONNECTIONS {
+            return Ok(());
+        }
+
+        // we don't know of any peers
+        if self.discovered.is_empty() {
+            return Err(P2pError::NoPeers);
+        }
+
+        let npeers = std::cmp::min(
+            self.discovered.len(),
+            MAX_ACTIVE_CONNECTIONS - self.peers.len(),
+        );
+
+        // TODO: improve peer selection
+        let mut peers = Vec::new();
+        let mut iter = self.discovered.iter();
+
+        for _ in 0..npeers {
+            let peer_info = iter.next().expect("Peer to exist");
+            let (ip4, ip6) = match peer_info.1 {
+                PeerAddrInfo::Raw { ip4, ip6 } => (ip4, ip6),
+            };
+            assert!(!ip4.is_empty() || !ip6.is_empty());
+
+            // TODO: let user specify their preference?
```

Comment:
Could you explain what this comment means?

Reply:
Mintlayer can fight the power and drive IPv6 adoption by preferring it over IPv4 :smiley: 

In all seriousness though, it's just something that we can decide for the user or the user can control if they so wish. It's extra work so might as well not give them the option. We know what's best for them.